atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/TracingKeyValueService.java

```diff
@@ -531,23 +585,32 @@ private void complete() {
         return future;
     }
 
-    private enum TableReferenceTagTranslator implements TagTranslator<TableReference> {
-        INSTANCE;
-
-        @Override
-        public <T> void translate(TagAdapter<T> adapter, T target, TableReference data) {
-            adapter.tag(
-                    target, "table", LoggingArgs.safeTableOrPlaceholder(data).toString());
-        }
-    }
-
-    private enum TableReferencesTagTranslator implements TagTranslator<Collection<TableReference>> {
-        INSTANCE;
-
-        @Override
-        public <T> void translate(TagAdapter<T> adapter, T target, Collection<TableReference> data) {
-            adapter.tag(
-                    target, "tables", LoggingArgs.safeTablesOrPlaceholder(data).toString());
-        }
+    /**
+     * Attach a detached span to the close of a closable iterator.
+     *
+     * Note: due to legacy code this span is not guaranteed to close. This will mean that some logs will be lost/missed.
+     */
+    @MustBeClosed
+    private static <V> ClosableIterator<V> attachDetachedSpanCompletion(
+            DetachedSpan detachedSpan, ClosableIterator<V> closableIterator, Consumer<TagConsumer> tagTranslator) {
+        // We'll possibly be completing in a different unrelated thread, so we have no chance to restore the original;
+        // this is a known limitation of the async trace statistics. We still want to clear the current statistics.
+        TraceStatistic original = TraceStatistics.getCurrentAndClear();
+        Thread originalThread = Thread.currentThread();
+
+        // As the trace executor will close out this span, we need to keep a reference to the current mutable statistics
+        TraceStatistic current = TraceStatistics.getReferenceToCurrent();
+
+        return ClosableIterators.appendOnClose(closableIterator, () -> {
+            detachedSpan.complete(FunctionalTagTranslator.INSTANCE, sink -> {
+                sink.statistics(current);
+                tagTranslator.accept(sink);
+
+                if (Thread.currentThread() == originalThread) {
+                    // Lucky, we can clean up the threadlocals
+                    TraceStatistics.getCopyAndRestoreOriginal(original);
+                }
+            });
+        });
```

Comment:
This is a bit unfortunate (with regards to not restoring the original), but yeah I don't really see a good way around that